[WIN] Stochastic bug: Minc2 data may load as np.longdouble in h5py < 2.10

These AppVeyor failures are strange and stochastic. The failures are consistent:

```
======================================================================
ERROR: autogenerated test from validate_data_deprecated
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python34\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\python34\lib\site-packages\nibabel\tests\test_api_validators.py", line 23, in meth
    validator(self, imaker, params)
  File "c:\python34\lib\site-packages\nibabel\tests\test_image_api.py", line 380, in validate_data_deprecated
    assert_data_similar(img._data, params)
  File "c:\python34\lib\site-packages\nibabel\tests\test_helpers.py", line 80, in assert_data_similar
    real_arr = np.asarray(arr)
  File "c:\python34\lib\site-packages\numpy\core\numeric.py", line 501, in asarray
    return array(a, dtype, copy=False, order=order)
  File "c:\python34\lib\site-packages\nibabel\minc1.py", line 261, in __array__
    return self.minc_file.get_scaled_data()
  File "c:\python34\lib\site-packages\nibabel\minc2.py", line 139, in get_scaled_data
    return self._normalize(raw_data, sliceobj)
  File "c:\python34\lib\site-packages\nibabel\minc1.py", line 189, in _normalize
    dmin, dmax = self._get_valid_range()
  File "c:\python34\lib\site-packages\nibabel\minc2.py", line 98, in _get_valid_range
    info = np.iinfo(ddt.type)
  File "c:\python34\lib\site-packages\numpy\core\getlimits.py", line 516, in __init__
    raise ValueError("Invalid integer data type.")
ValueError: Invalid integer data type.
======================================================================
ERROR: nibabel.tests.test_minc2.TestMinc2File.test_array_proxy_slicing
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python34\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\python34\lib\site-packages\nibabel\tests\test_minc1.py", line 200, in test_array_proxy_slicing
    arr = img.get_data()
  File "c:\python34\lib\site-packages\nibabel\dataobj_images.py", line 202, in get_data
    data = np.asanyarray(self._dataobj)
  File "c:\python34\lib\site-packages\numpy\core\numeric.py", line 553, in asanyarray
    return array(a, dtype, copy=False, order=order, subok=True)
  File "c:\python34\lib\site-packages\nibabel\minc1.py", line 261, in __array__
    return self.minc_file.get_scaled_data()
  File "c:\python34\lib\site-packages\nibabel\minc2.py", line 139, in get_scaled_data
    return self._normalize(raw_data, sliceobj)
  File "c:\python34\lib\site-packages\nibabel\minc1.py", line 189, in _normalize
    dmin, dmax = self._get_valid_range()
  File "c:\python34\lib\site-packages\nibabel\minc2.py", line 98, in _get_valid_range
    info = np.iinfo(ddt.type)
  File "c:\python34\lib\site-packages\numpy\core\getlimits.py", line 516, in __init__
    raise ValueError("Invalid integer data type.")
ValueError: Invalid integer data type.
======================================================================
ERROR: nibabel.tests.test_minc2.TestMinc2File.test_load
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python34\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\python34\lib\site-packages\nibabel\tests\test_minc1.py", line 187, in test_load
    data = img.get_data()
  File "c:\python34\lib\site-packages\nibabel\dataobj_images.py", line 202, in get_data
    data = np.asanyarray(self._dataobj)
  File "c:\python34\lib\site-packages\numpy\core\numeric.py", line 553, in asanyarray
    return array(a, dtype, copy=False, order=order, subok=True)
  File "c:\python34\lib\site-packages\nibabel\minc1.py", line 261, in __array__
    return self.minc_file.get_scaled_data()
  File "c:\python34\lib\site-packages\nibabel\minc2.py", line 139, in get_scaled_data
    return self._normalize(raw_data, sliceobj)
  File "c:\python34\lib\site-packages\nibabel\minc1.py", line 189, in _normalize
    dmin, dmax = self._get_valid_range()
  File "c:\python34\lib\site-packages\nibabel\minc2.py", line 98, in _get_valid_range
    info = np.iinfo(ddt.type)
  File "c:\python34\lib\site-packages\numpy\core\getlimits.py", line 516, in __init__
    raise ValueError("Invalid integer data type.")
ValueError: Invalid integer data type.
======================================================================
ERROR: nibabel.tests.test_minc2.TestMinc2File.test_mincfile_slicing
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python34\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\python34\lib\site-packages\nibabel\tests\test_minc1.py", line 171, in test_mincfile_slicing
    data = mnc.get_scaled_data()
  File "c:\python34\lib\site-packages\nibabel\minc2.py", line 139, in get_scaled_data
    return self._normalize(raw_data, sliceobj)
  File "c:\python34\lib\site-packages\nibabel\minc1.py", line 189, in _normalize
    dmin, dmax = self._get_valid_range()
  File "c:\python34\lib\site-packages\nibabel\minc2.py", line 98, in _get_valid_range
    info = np.iinfo(ddt.type)
  File "c:\python34\lib\site-packages\numpy\core\getlimits.py", line 516, in __init__
    raise ValueError("Invalid integer data type.")
ValueError: Invalid integer data type.
======================================================================
ERROR: nibabel.tests.test_processing.test_spatial_axes_check
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python34\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\python34\lib\site-packages\numpy\testing\_private\decorators.py", line 155, in skipper_func
    return f(*args, **kwargs)
  File "c:\python34\lib\site-packages\nibabel\tests\test_processing.py", line 366, in test_spatial_axes_check
    out = resample_from_to(img, img, mode='nearest')
  File "c:\python34\lib\site-packages\nibabel\processing.py", line 180, in resample_from_to
    cval=cval)
  File "c:\python34\lib\site-packages\scipy\ndimage\interpolation.py", line 416, in affine_transform
    filtered = spline_filter(input, order, output=numpy.float64)
  File "c:\python34\lib\site-packages\scipy\ndimage\interpolation.py", line 113, in spline_filter
    spline_filter1d(input, order, axis, output=output)
  File "c:\python34\lib\site-packages\scipy\ndimage\interpolation.py", line 82, in spline_filter1d
    _nd_image.spline_filter1d(input, order, axis, output)
RuntimeError: array type dtype('float64') not supported
======================================================================
FAIL: autogenerated test from validate_data_interface
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python34\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\python34\lib\site-packages\nibabel\tests\test_api_validators.py", line 23, in meth
    validator(self, imaker, params)
  File "c:\python34\lib\site-packages\nibabel\tests\test_image_api.py", line 208, in validate_data_interface
    self._check_proxy_interface(imaker, meth_name)
  File "c:\python34\lib\site-packages\nibabel\tests\test_image_api.py", line 253, in _check_proxy_interface
    assert_true(data is data_again)
AssertionError: False is not true
======================================================================
FAIL: autogenerated test from validate_dtype
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python34\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\python34\lib\site-packages\nibabel\tests\test_api_validators.py", line 23, in meth
    validator(self, imaker, params)
  File "c:\python34\lib\site-packages\nibabel\tests\test_image_api.py", line 178, in validate_dtype
    assert_equal(img.get_data_dtype().type, params['dtype'])
AssertionError: <class 'numpy.float64'> != <class 'numpy.float64'>
======================================================================
FAIL: nibabel.tests.test_minc2.TestMinc2File.test_mincfile
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python34\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\python34\lib\site-packages\nibabel\tests\test_minc1.py", line 159, in test_mincfile
    assert_equal(mnc.get_data_dtype().type, tp['dtype'])
AssertionError: <class 'numpy.float64'> != <class 'numpy.float64'>
```

And I've seen them show up in Python 3.4, Python 3.4-x64, Python 3.5 and Python 3.5-x64, but in none of these cases do they consistently fail.

Example builds:

* https://ci.appveyor.com/project/nipy/nibabel/build/1.0.476
* https://ci.appveyor.com/project/nipy/nibabel/build/1.0.477

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -5,9 +5,19 @@ environment:
   DEPENDS: numpy scipy matplotlib h5py pydicom
 
   matrix:
+    - PYTHON: C:\Python27
+    - PYTHON: C:\Python27-x64
+    - PYTHON: C:\Python34
+    - PYTHON: C:\Python34-x64
+    - PYTHON: C:\Python35
+    - PYTHON: C:\Python35-x64
     - PYTHON: C:\Python35-x64
       PYTHONHASHSEED: 283137131
       DEPENDS: "h5py==2.9.0"
+    - PYTHON: C:\Python36
+    - PYTHON: C:\Python36-x64
+    - PYTHON: C:\Python37
+    - PYTHON: C:\Python37-x64
 
 install:
   # Prepend newly installed Python to the PATH of this build (this cannot be

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -8,7 +8,9 @@ environment:
     - PYTHON: C:\Python27
     - PYTHON: C:\Python27-x64
     - PYTHON: C:\Python34
+      DEPENDS: --prefer-binary numpy scipy matplotlib h5py pydicom
     - PYTHON: C:\Python34-x64
+      DEPENDS: --prefer-binary numpy scipy matplotlib h5py pydicom
     - PYTHON: C:\Python35
     - PYTHON: C:\Python35-x64
     - PYTHON: C:\Python35-x64

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -5,19 +5,9 @@ environment:
   DEPENDS: numpy scipy matplotlib h5py pydicom
 
   matrix:
-    - PYTHON: C:\Python27
-    - PYTHON: C:\Python27-x64
-    - PYTHON: C:\Python34
-    - PYTHON: C:\Python34-x64
-    - PYTHON: C:\Python35
-    - PYTHON: C:\Python35-x64
     - PYTHON: C:\Python35-x64
       PYTHONHASHSEED: 283137131
       DEPENDS: "h5py==2.9.0"
-    - PYTHON: C:\Python36
-    - PYTHON: C:\Python36-x64
-    - PYTHON: C:\Python37
-    - PYTHON: C:\Python37-x64
 
 install:
   # Prepend newly installed Python to the PATH of this build (this cannot be

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -2,6 +2,7 @@
 # CI on Windows via appveyor
 
 environment:
+  DEPENDS: numpy scipy matplotlib h5py pydicom
 
   matrix:
     - PYTHON: C:\Python27
@@ -10,6 +11,9 @@ environment:
     - PYTHON: C:\Python34-x64
     - PYTHON: C:\Python35
     - PYTHON: C:\Python35-x64
+    - PYTHON: C:\Python35-x64
+      PYTHONHASHSEED: 283137131
+      DEPENDS: "h5py==2.9.0"
     - PYTHON: C:\Python36
     - PYTHON: C:\Python36-x64
     - PYTHON: C:\Python37
@@ -28,7 +32,7 @@ install:
 
   # Install the dependencies of the project.
   - pip install --upgrade pip setuptools>=27.0 wheel
-  - pip install numpy scipy matplotlib h5py pydicom
+  - pip install %DEPENDS%
   - pip install nose mock coverage codecov
   - pip install .
   - SET NIBABEL_DATA_DIR=%CD%\nibabel-data

--- a/nibabel/_h5py_compat.py
+++ b/nibabel/_h5py_compat.py
@@ -1,0 +1,12 @@
+import sys
+import os
+from .optpkg import optional_package
+
+# PY35: A bug affected Windows installations of h5py in Python3 versions <3.6
+# due to random dictionary ordering, causing float64 data arrays to sometimes be
+# loaded as longdouble (also 64 bit on Windows). This caused stochastic failures
+# to correctly handle data caches, and possibly other subtle bugs we never
+# caught. This was fixed in h5py 2.10.
+# Please see https://github.com/nipy/nibabel/issues/665 for details.
+min_h5py = '2.10' if os.name == 'nt' and (3,) <= sys.version_info < (3, 6) else None
+h5py, have_h5py, setup_module = optional_package('h5py', min_version=min_h5py)

--- a/nibabel/minc1.py
+++ b/nibabel/minc1.py
@@ -173,7 +173,7 @@ class Minc1File(object):
             applied to `data`
         """
         ddt = self.get_data_dtype()
-        if ddt.type in np.sctypes['float']:
+        if np.issubdtype(ddt.type, np.floating):
             return data
         image_max = self._image_max
         image_min = self._image_min

--- a/nibabel/minc2.py
+++ b/nibabel/minc2.py
@@ -25,21 +25,10 @@ and compare against command line output of::
 
     mincstats my_funny.mnc
 """
-import sys
-import os
 import numpy as np
 
 from .keywordonly import kw_only_meth
-from .optpkg import optional_package
-
-# PY35: A bug affected Windows installations of h5py in Python3 versions <3.6
-# due to random dictionary ordering, causing float64 data arrays to sometimes be
-# loaded as longdouble (also 64 bit on Windows). This caused stochastic failures
-# to correctly handle data caches, and possibly other subtle bugs we never
-# caught. This was fixed in h5py 2.10.
-# Please see https://github.com/nipy/nibabel/issues/665 for details.
-min_h5py = '2.10' if os.name == 'nt' and (3,) <= sys.version_info < (3, 6) else None
-h5py, have_h5py, setup_module = optional_package('h5py', min_version=min_h5py)
+from ._h5py_compat import h5py
 
 from .minc1 import Minc1File, MincHeader, Minc1Image, MincError
 

--- a/nibabel/minc2.py
+++ b/nibabel/minc2.py
@@ -25,11 +25,21 @@ and compare against command line output of::
 
     mincstats my_funny.mnc
 """
+import sys
+import os
 import numpy as np
 
 from .keywordonly import kw_only_meth
 from .optpkg import optional_package
-h5py, have_h5py, setup_module = optional_package('h5py')
+
+# PY35: A bug affected Windows installations of h5py in Python3 versions <3.6
+# due to random dictionary ordering, causing float64 data arrays to sometimes be
+# loaded as longdouble (also 64 bit on Windows). This caused stochastic failures
+# to correctly handle data caches, and possibly other subtle bugs we never
+# caught. This was fixed in h5py 2.10.
+# Please see https://github.com/nipy/nibabel/issues/665 for details.
+min_h5py = '2.10' if os.name == 'nt' and (3,) <= sys.version_info < (3, 6) else None
+h5py, have_h5py, setup_module = optional_package('h5py', min_version=min_h5py)
 
 from .minc1 import Minc1File, MincHeader, Minc1Image, MincError
 

--- a/nibabel/tests/test_h5py_compat.py
+++ b/nibabel/tests/test_h5py_compat.py
@@ -1,0 +1,44 @@
+"""
+These tests are almost certainly overkill, but serve to verify that
+the behavior of _h5py_compat is pass-through in all but a small set of
+well-defined cases
+"""
+import sys
+import os
+from distutils.version import LooseVersion
+import numpy as np
+
+from ..optpkg import optional_package
+from .. import _h5py_compat as compat
+from ..testing import assert_equal, assert_true, assert_false, assert_not_equal
+
+h5py, have_h5py, _ = optional_package('h5py')
+
+
+def test_optpkg_equivalence():
+    # No effect on Linux/OSX
+    if os.name == 'posix':
+        assert_equal(have_h5py, compat.have_h5py)
+    # No effect on Python 2.7 or 3.6+
+    if sys.version_info >= (3, 6) or sys.version_info < (3,):
+        assert_equal(have_h5py, compat.have_h5py)
+    # Available in a strict subset of cases
+    if not have_h5py:
+        assert_false(compat.have_h5py)
+    # Available when version is high enough
+    elif LooseVersion(h5py.__version__) >= '2.10':
+        assert_true(compat.have_h5py)
+
+
+def test_disabled_h5py_cases():
+    # On mismatch
+    if have_h5py and not compat.have_h5py:
+        # Recapitulate min_h5py conditions from _h5py_compat
+        assert_equal(os.name, 'nt')
+        assert_true((3,) <= sys.version_info < (3, 6))
+        assert_true(LooseVersion(h5py.__version__) < '2.10')
+        # Verify that the root cause is present
+        # If any tests fail, they will likely be these, so they may be
+        # ill-advised...
+        assert_equal(str(np.longdouble), str(np.float64))
+        assert_not_equal(np.longdouble, np.float64)

--- a/nibabel/tests/test_image_api.py
+++ b/nibabel/tests/test_image_api.py
@@ -34,7 +34,7 @@ import numpy as np
 
 from ..optpkg import optional_package
 _, have_scipy, _ = optional_package('scipy')
-from ..minc2 import have_h5py
+from .._h5py_compat import have_h5py
 
 from .. import (AnalyzeImage, Spm99AnalyzeImage, Spm2AnalyzeImage,
                 Nifti1Pair, Nifti1Image, Nifti2Pair, Nifti2Image,

--- a/nibabel/tests/test_image_api.py
+++ b/nibabel/tests/test_image_api.py
@@ -34,7 +34,7 @@ import numpy as np
 
 from ..optpkg import optional_package
 _, have_scipy, _ = optional_package('scipy')
-_, have_h5py, _ = optional_package('h5py')
+from ..minc2 import have_h5py
 
 from .. import (AnalyzeImage, Spm99AnalyzeImage, Spm2AnalyzeImage,
                 Nifti1Pair, Nifti1Image, Nifti2Pair, Nifti2Image,

--- a/nibabel/tests/test_imageclasses.py
+++ b/nibabel/tests/test_imageclasses.py
@@ -10,7 +10,7 @@ import nibabel as nib
 from nibabel.analyze import AnalyzeImage
 from nibabel.nifti1 import Nifti1Image
 from nibabel.nifti2 import Nifti2Image
-from ..minc2 import have_h5py
+from .._h5py_compat import have_h5py
 
 from nibabel import imageclasses
 from nibabel.imageclasses import spatial_axes_first, class_map, ext_map

--- a/nibabel/tests/test_imageclasses.py
+++ b/nibabel/tests/test_imageclasses.py
@@ -6,12 +6,11 @@ import warnings
 
 import numpy as np
 
-from nibabel.optpkg import optional_package
-
 import nibabel as nib
 from nibabel.analyze import AnalyzeImage
 from nibabel.nifti1 import Nifti1Image
 from nibabel.nifti2 import Nifti2Image
+from ..minc2 import have_h5py
 
 from nibabel import imageclasses
 from nibabel.imageclasses import spatial_axes_first, class_map, ext_map
@@ -22,8 +21,6 @@ from nibabel.testing import clear_and_catch_warnings
 
 
 DATA_DIR = pjoin(dirname(__file__), 'data')
-
-have_h5py = optional_package('h5py')[1]
 
 MINC_3DS = ('minc1_1_scale.mnc',)
 MINC_4DS = ('minc1_4d.mnc',)

--- a/nibabel/tests/test_minc2.py
+++ b/nibabel/tests/test_minc2.py
@@ -13,7 +13,8 @@ from os.path import join as pjoin
 import numpy as np
 
 from .. import minc2
-from ..minc2 import Minc2File, Minc2Image, h5py, have_h5py, setup_module
+from ..minc2 import Minc2File, Minc2Image
+from .._h5py_compat import h5py, have_h5py, setup_module
 
 from nose.tools import (assert_true, assert_equal, assert_false, assert_raises)
 

--- a/nibabel/tests/test_minc2.py
+++ b/nibabel/tests/test_minc2.py
@@ -12,12 +12,8 @@ from os.path import join as pjoin
 
 import numpy as np
 
-from ..optpkg import optional_package
-
-h5py, have_h5py, setup_module = optional_package('h5py')
-
 from .. import minc2
-from ..minc2 import Minc2File, Minc2Image
+from ..minc2 import Minc2File, Minc2Image, h5py, have_h5py, setup_module
 
 from nose.tools import (assert_true, assert_equal, assert_false, assert_raises)
 

--- a/nibabel/tests/test_minc2_data.py
+++ b/nibabel/tests/test_minc2_data.py
@@ -15,7 +15,7 @@ from os.path import join as pjoin
 
 import numpy as np
 
-from ..minc2 import h5py, have_h5py, setup_module
+from .._h5py_compat import h5py, have_h5py, setup_module
 
 from .nibabel_data import get_nibabel_data, needs_nibabel_data
 from .. import load as top_load, Nifti1Image

--- a/nibabel/tests/test_minc2_data.py
+++ b/nibabel/tests/test_minc2_data.py
@@ -15,9 +15,7 @@ from os.path import join as pjoin
 
 import numpy as np
 
-from nibabel.optpkg import optional_package
-
-h5py, have_h5py, setup_module = optional_package('h5py')
+from ..minc2 import h5py, have_h5py, setup_module
 
 from .nibabel_data import get_nibabel_data, needs_nibabel_data
 from .. import load as top_load, Nifti1Image

--- a/nibabel/tests/test_proxy_api.py
+++ b/nibabel/tests/test_proxy_api.py
@@ -46,7 +46,7 @@ from ..freesurfer.mghformat import MGHHeader
 from .. import minc1
 from ..externals.netcdf import netcdf_file
 from .. import minc2
-from ..minc2 import h5py, have_h5py
+from .._h5py_compat import h5py, have_h5py
 from .. import ecat
 from .. import parrec
 

--- a/nibabel/tests/test_proxy_api.py
+++ b/nibabel/tests/test_proxy_api.py
@@ -46,8 +46,7 @@ from ..freesurfer.mghformat import MGHHeader
 from .. import minc1
 from ..externals.netcdf import netcdf_file
 from .. import minc2
-from ..optpkg import optional_package
-h5py, have_h5py, _ = optional_package('h5py')
+from ..minc2 import h5py, have_h5py
 from .. import ecat
 from .. import parrec
 

